src/cargo/core/resolver/types.rs

```diff
@@ -324,22 +324,24 @@ pub type ConflictMap = BTreeMap<PackageId, ConflictReason>;
 
 pub struct RcVecIter<T> {
     vec: Rc<Vec<T>>,
-    rest: Range<usize>,
+    offset: usize,
 }
 
 impl<T> RcVecIter<T> {
     pub fn new(vec: Rc<Vec<T>>) -> RcVecIter<T> {
-        RcVecIter {
-            rest: 0..vec.len(),
-            vec,
-        }
+        RcVecIter { vec, offset: 0 }
+    }
+
+    pub fn peek(&self) -> Option<&T> {
+        self.vec.get(self.offset)
     }
 
-    fn peek(&self) -> Option<(usize, &T)> {
-        self.rest
-            .clone()
-            .next()
-            .and_then(|i| self.vec.get(i).map(|val| (i, &*val)))
+    pub fn remaining(&self) -> impl Iterator<Item = &T> + '_ {
+        self.vec.get(self.offset..).into_iter().flatten()
+    }
+
+    pub fn iter_by_ref(&mut self) -> impl Iterator<Item = &T> + '_ {
+        (self.vec.get(self.offset..).into_iter().flatten()).inspect(|_| self.offset += 1)
```

Comment:
We are relying on some spooky action keeping the iteration counter inside the range `self.offset..` to the same value as the `self.offset`. Put a different way, using `inspect` for its side effects makes my skin crawl.

The best I've got for an alternative way to spell it is:
```rust
        std::iter::from_fn(|| {
            let out = self.vec.get(self.offset);
            if out.is_some() {
                self.offset += 1;
            }
            out
        })
```

Although that may end up being slower because of the loss of `ExactSizeIterator`.

If we can't find a better way of spelling it, we should at least put a comment within the `inspect` explain why it's critical for correctness.